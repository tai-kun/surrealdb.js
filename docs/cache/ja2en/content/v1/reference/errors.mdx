---
title: Errors
slug: en/v1/reference/errors
---

## SurrealError

`SurrealError` is the base class that most of the error objects explicitly thrown by this SDK inherit from. This class inherits from JavaScript's `Error` class. This error object is rarely thrown alone. It is usually a specialized error object that inherits from it and has a special meaning. It is recommended to use this error object to easily classify errors that have been caught using the `try-catch` construct as errors explicitly thrown by this SDK.

`.name: "SurrealError"`
:   Name of the error.

`.message`
:   Error message.

`.stack`
:   Stack trace may be recorded.

`.cause`
:   The cause or context of the error may be set.

## SurrealTypeError

Inherits from: `SurrealError`

This error is thrown when an input value or other verification fails.

`.name: "SurrealTypeError"`
:   Name of the error.

`.expected`
:   Expected data type.

`.actual`
:   String representation of the actual value.

## SurrealAggregateError

Inherits from: `SurrealError`

This is an error that aggregates multiple errors or error messages.

`.name: "SurrealAggregateError"`
:   Name of the error.

`.cause: unknown[]`
:   Each element of the array contains errors, error messages, etc.

## CircularReferenceError

Inherits from: `SurrealError`

This error is thrown when a circular reference is detected. It is mainly thrown when converting a JavaScript value to another format, and the same object as the parent object is detected.

`.name: "CircularReferenceError"`
:   Name of the error.

For example, this error is thrown in the following case:

```ts
import { toSurql } from "@tai-kun/surrealdb/utils";

const a = {};
a.a = a;

console.log(a); // <ref *1> { a: [Circular *1] }

toSurql(a); // throws CircularReferenceError
```

## NumberRangeError

Inherits from: `SurrealError`

This error is thrown when a number takes a value outside the range.

`.name: "NumberRangeError"`
:   Name of the error.

`.range`
:   Expected range.

`.actual`
:   Actual value.

`.integer`
:   If `true`, it expects an integer.

## UnsupportedRuntimeError

Inherits from: `SurrealError`

This error is thrown when it is determined that an unsupported runtime is being used.

`.name: "UnsupportedRuntimeError"`
:   Name of the error.

## UnreachableError

Inherits from: `SurrealError`

This error is thrown when unreachable code is reached. If this error is thrown, you are probably encountering a bug in this SDK.

`.name: "UnreachableError"`
:   Name of the error.

## CircularEngineReferenceError

Inherits from: `CircularReferenceError`

This error is thrown when a circular reference occurs between engines when connecting to a database.

`.name: "CircularEngineReferenceError"`
:   Name of the error.

`.seen: string[]`
:   List of protocol names passed through until a circular reference occurs.

For example, this error is thrown in the following case:

```ts
const { Surreal } = initSurreal({
  engines: {
    http: "https",
    https: "http",
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws CircularEngineReferenceError: Circular engine reference: http,https
```

In the above example, the https protocol engine is set to use the engine set to http, but the http protocol engine tries to use the engine set to https, resulting in a circular reference error.

## EngineNotFoundError

Inherits from: `SurrealError`

This error is thrown when you try to connect with a protocol that has no engine set up.

`.name: "EngineNotFoundError"`
:   Name of the error.

`.protocol: string`
:   Protocol name of the connection destination.

For example, this error is thrown in the following case:

```ts
const { Surreal } = initSurreal({
  engines: {
    http: config => new HttpEngine(config),
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws EngineNotFoundError: No https protocol engine found.
```

## ConnectionConflictError

Inherits from: `SurrealError`

This error is thrown when a client attempts to connect to multiple endpoints simultaneously.

`.name: "ConnectionConflictError"`
:   Name of the error.

`.endpoint1: string`
:   One of the endpoints.

`.endpoint2: string`
:   The other endpoint.

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.connect("http://localhost:11298"); // throws ConnectionConflictError: Connection conflict between http://localhost:8000/rpc and http://localhost:11298/rpc.
```

The `.connect` method adds `/rpc` to the end of the URL if the path of the passed URL does not end with `/rpc`. Therefore, if the URL path of one endpoint ends with `/rpc`, it may not result in an error, contrary to its appearance:

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.connect("http://localhost:8000/rpc"); // OK!
```

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000/rpc");
await db.connect("http://localhost:8000"); // OK!
```

## NamespaceConflictError

Inherits from: `SurrealError`

This error is thrown when the SDK attempts to switch to multiple different namespaces simultaneously.

`.name: "NamespaceConflictError"`
:   Name of the error.

`.namespace1: string`
:   One of the namespaces.

`.namespace2: string`
:   The other namespace.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.signin({ user: "root", pass: "root" });
await Promise.all([
  db.use({ namespace: "foo" }),
  db.use({ namespace: "bar" }),
]);
```

## DatabaseConflictError

Inherits from: `SurrealError`

This error is thrown when the SDK attempts to switch to multiple different databases simultaneously.

`.name: "DatabaseConflictError"`
:   Name of the error.

`.database1: string`
:   One of the databases.

`.database2: string`
:   The other database.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.signin({ user: "root", pass: "root" });
await Promise.all([
  db.use({ namespace: "baz" database: "foo" }),
  db.use({ namespace: "baz" database: "bar" }),
]);
```

## MissingNamespaceError

Inherits from: `SurrealError`

This error is thrown when a namespace is not selected before a database is selected. Or it is also thrown when you try to make the namespace unselected while the database is selected.

`.name: "MissingNamespaceError"`
:   Name of the error.

`.database: string`
:   Database name.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.use({ database: "example" }); // throws MissingNamespaceError: The namespace must be specified before the database.
```

## RpcResponseError

Inherits from: `SurrealError`

This error is thrown when the RPC response indicates an error. There is no problem with communication or decoding the response body due to the connected protocol, but it means that SurrealDB cannot process the RPC request.

`.name: "RpcResponseError"`
:   Name of the error.

`.id?: string`
:   ID to identify the RPC request. The ID always starts with `<method name>`.

`.code: number`
:   It is not explicitly documented in SurrealDB, but it is probably a JSON-RPC error code.

## QueryFailedError

Inherits from: `SurrealAggregateError`

This error is thrown when a query fails.

`.name: "QueryFailedError"`
:   Name of the error.

`.cause: string[]`
:   List of errors.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.query("OUTPUT 'Hello'"); // throws QueryFailedError: Query failed with 1 error(s)
```

## Disconnected

Inherits from: `SurrealError`

This error indicates that the connection has been forcibly terminated.

`.name: "Disconnected"`
:   Name of the error.

For example, you get this error in the following case:

```ts
const db = new Surreal();

db.on("<event_name>", ({ signal }) => {
  return new Promise((resolve, reject) => {
    signal.addEventListener("abort", function() {
      console.error(this.reason); // Disconnected
      reject();
    })

    // ...
  });
});

await db.disconnect({ force: true });
```

## EngineError

Inherits from: `SurrealError`

This indicates that the error originated from the engine. It is dispatched via an event listener.

`.name: "EngineError"`
:   Name of the error.

`.fatal: boolean | undefined`
:   Indicates whether this error is fatal.

## HttpEngineError

Inherits from: `EngineError`

This indicates that the error originated from the HTTP engine. Currently, it is only defined and not used.

`.name: "HttpEngineError"`
:   Name of the error.

## WebSocketEngineError

Inherits from: `EngineError`

This indicates that the error originated from the WebSocket engine. It is used when parsing the response body fails before finding the ID that identifies the RPC request, when an error event is received from the WebSocket object, when the WebSocket connection is disconnected, etc.

`.name: "WebSocketEngineError"`
:   Name of the error.

`.code: number`
:   One of the following status codes:
    - `1000`~`1015`: [MDN Reference](https://developer.mozilla.org/docs/Web/API/CloseEvent/code)
    - `3150`: Indicates that an "error" event was received from the `WebSocket`[^1] instance.
    - `3151`: Indicates that an error occurred in the "open" event handler registered by the WebSocket engine. Most likely, `StateTransitionError` occurred during the transition from the connecting state to the connected state.
    - `3152`: Indicates that an error occurred in the "message" event handler registered by the WebSocket engine. There are many possible causes, but it is most likely that `ResponseError` or `RpcResponseError` was thrown in the event handler. If you trace back, it is likely that an invalid RPC parameter was requested or that an incorrect grammar SurrealQL custom query was requested in the `.query()` method. In that case, the RPC caller method will timeout and fail (by default) in 5 seconds.
    - `3153`: Indicates that the ping transmission failed. It may be a temporary error, but if you continue to receive it, it is possible that the connection is not being maintained.
    - `3154`: Indicates that an error occurred in the "close" event handler registered by the WebSocket engine. Most likely, `StateTransitionError` occurred during the transition from the disconnecting state to the disconnected state.
:   Note that the following status codes are excluded from the above. This SDK does not treat them as errors.
    - `1000`: Normally disconnected.
    - `1001`: Early disconnection is common.
    - `1004`: Reserved.
    - `1005`: Reserved.
    - `1006`: Reserved.
    - `1015`: Reserved.

^1: `WebSocket` can be a class defined in a global variable by the runtime or a class in the `ws` SDK.

## StateTransitionError

Inherits from: `SurrealAggregateError`

This error is thrown when the execution of an event listener fails during a state transition.

`.name: "StateTransitionError"`
:   Name of the error.

`.from: number`
:   State at the start of the state transition.

`.to: number`
:   Target state of the transition.

`.fallback: number`
:   Alternative transition destination if state transition fails. If it is the same value as `.to`, it means that it was forcibly transitioned.

For example, this error is thrown in the following case:

```ts
import { Surreal } from "@tai-kun/surrealdb";
import { CONNECTING } from "@tai-kun/surrealdb/engine";

await using db = new Surreal();

db.on(CONNECTING, () => {
  throw new Error("Don't let me connect !!!")
});

await db.connect("http://localhost:8000"); // throws ...
// StateTransitionError: The transition from `3` to `0` failed, falling back to `3`.
//     at <stack trace ...> {
//   cause: [
//     Error: Don't let me connect !!!
//         at <stack trace ...>
//   ],
//   from: 3,
//   to: 0,
//   fallback: 3
// }
```

## ConnectionUnavailableError

Inherits from: `SurrealError`

This error is thrown when you try to send an RPC request in an unconnected state.

`.name: "ConnectionUnavailableError"`
:   Name of the error.

## ResponseError

Inherits from: `SurrealError`

This error is thrown when a response cannot be parsed as a PRC response. It is different from `RpcResponseError`.

`.name: "ResponseError"`
:   Name of the error.

## CborError

Inherits from: `SurrealError`

This is the class that all errors related to CBOR explicitly thrown by `@tai-kun/surrealdb/cbor` inherit from. It is never thrown directly.

`.name: "CborError"`
:   Name of the error.

## CborWellFormednessError

Inherits from: `CborError`

This is the class that all errors related to CBOR decoding explicitly thrown by `@tai-kun/surrealdb/cbor` inherit from. It is never thrown directly.

See "Well-Formedness Errors and Examples" in Appendix F of RFC8949 [https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-](https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-).

`.name: "CborWellFormednessError"`
:   Name of the error.

## CborTooMuchDataError

Inherits from: `CborWellFormednessError`

Indicates that there are unconsumed input bytes left.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2)

`.name: "CborTooMuchDataError"`
:   Name of the error.

## CborTooLittleDataError

Inherits from: `CborWellFormednessError`

Indicates that the input bytes are incomplete.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4)

`.name: "CborTooLittleDataError"`
:   Name of the error.

## CborSyntaxError

Inherits from: `CborWellFormednessError`

Indicates that the input bytes do not match the CBOR encoding format.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6)

`.name: "CborSyntaxError"`
:   Name of the error.

## CborMaxDepthReachedError

Inherits from: `CborError`

This error is thrown when the depth of the JavaScript object reaches the maximum value during CBOR encoding or decoding. The depth increases by 1 each time you enter an object or array.

`.name: "CborMaxDepthReachedError"`
:   Name of the error.

`.maxDepth: number`
:   Maximum depth.

## CborUnsafeMapKeyError

Inherits from: `CborError`

This error is thrown when an unsafe map key is found during CBOR encoding or decoding.

`.name: "CborUnsafeMapKeyError"`
:   Name of the error.

`.key: unknown`
:   The map key that was determined to be unsafe.
