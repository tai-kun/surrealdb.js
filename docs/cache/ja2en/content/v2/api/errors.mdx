---
title: Errors
slug: en/v2/api/errors
---

## SurrealError

`SurrealError` is the base class that most error objects explicitly thrown by this SDK inherit from. This class inherits from JavaScript's `Error` class. This error object is rarely thrown on its own, and it is generally thrown by its inheriting error objects with specific meanings. It is recommended to use it to easily classify errors caught in `try-catch` blocks as errors explicitly thrown by this SDK.

`.name: "SurrealError"`
:   The error name.

`.message`
:   The error message.

`.stack`
:   Stack trace may be recorded.

`.cause`
:   The cause or context of the error may be set.

## SurrealTypeError

Inherits from: `SurrealError`

This error is thrown when an input value or other verification fails.

`.name: "SurrealTypeError"`
:   The error name.

`.expected`
:   The expected data format.

`.actual`
:   The string representation of the actual value.

## SurrealAggregateError

Inherits from: `SurrealError`

This is an error that encapsulates multiple errors or error messages.

`.name: "SurrealAggregateError"`
:   The error name.

`.cause: unknown[]`
:   Each element of the array contains an error or error message.

## CircularReferenceError

Inherits from: `SurrealError`

This error is thrown when a circular reference is detected. It is mainly thrown when converting JavaScript values to another format, when a parent object is detected to be the same as the object itself.

`.name: "CircularReferenceError"`
:   The error name.

For example, this error is thrown in the following case:

```ts
import { toSurql } from "@tai-kun/surrealdb/utils";

const a = {};
a.a = a;

console.log(a); // <ref *1> { a: [Circular *1] }

toSurql(a); // throws CircularReferenceError
```

## NumberRangeError

Inherits from: `SurrealError`

This error is thrown when a number is out of range.

`.name: "NumberRangeError"`
:   The error name.

`.range`
:   The expected range.

`.actual`
:   The actual value.

`.integer`
:   If `true`, an integer is expected.

## UnsupportedRuntimeError

Inherits from: `SurrealError`

This error is thrown when it is determined that an unsupported runtime is being used.

`.name: "UnsupportedRuntimeError"`
:   The error name.

## UnreachableError

Inherits from: `SurrealError`

This error is thrown when unreachable code is reached. If this error is thrown, you have likely encountered a bug in this SDK.

`.name: "UnreachableError"`
:   The error name.

## CircularEngineReferenceError

Inherits from: `CircularReferenceError`

This error is thrown when a circular reference occurs between engines when connecting to the database.

`.name: "CircularEngineReferenceError"`
:   The error name.

`.seen: string[]`
:   List of protocol names passed through before the circular reference occurred.

For example, this error is thrown in the following case:

```ts
const { Surreal } = initSurreal({
  engines: {
    http: "https",
    https: "http",
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws CircularEngineReferenceError: Circular engine reference: http,https
```

In the above example, although the https protocol engine indicates that it will use the engine set for http, the http protocol engine tries to use the engine set for https, resulting in a circular reference error.

## EngineNotFoundError

Inherits from: `SurrealError`

This error is thrown when attempting to connect with a protocol that does not have an engine configured.

`.name: "EngineNotFoundError"`
:   The error name.

`.protocol: string`
:   The protocol name of the connection target.

For example, this error is thrown in the following case:

```ts
const { Surreal } = initSurreal({
  engines: {
    http: config => new HttpEngine(config),
  },
  // ...
});

await using db = new Surreal();
await db.connect("https://localhost:8000"); // throws EngineNotFoundError: No https protocol engine found.
```

## ConnectionConflictError

Inherits from: `SurrealError`

This error is thrown when a client attempts to connect to multiple endpoints simultaneously.

`.name: "ConnectionConflictError"`
:   The error name.

`.endpoint1: string`
:   One of the endpoints.

`.endpoint2: string`
:   The other endpoint.

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.connect("http://localhost:11298"); // throws ConnectionConflictError: Connection conflict between http://localhost:8000/rpc and http://localhost:11298/rpc.
```

The `.connect` method adds `/rpc` to the end of the URL path if it does not end with it. Therefore, if the URL path of one of the endpoints ends with `/rpc`, it may not result in an error, contrary to appearances:

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.connect("http://localhost:8000/rpc"); // OK!
```

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000/rpc");
await db.connect("http://localhost:8000"); // OK!
```

## NamespaceConflictError

Inherits from: `SurrealError`

This error is thrown when the SDK attempts to switch to multiple different namespaces simultaneously.

`.name: "NamespaceConflictError"`
:   The error name.

`.namespace1: string`
:   One of the namespaces.

`.namespace2: string`
:   The other namespace.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.signin({ user: "root", pass: "root" });
await Promise.all([
  db.use({ namespace: "foo" }),
  db.use({ namespace: "bar" }),
]);
```

## DatabaseConflictError

Inherits from: `SurrealError`

This error is thrown when the SDK attempts to switch to multiple different databases simultaneously.

`.name: "DatabaseConflictError"`
:   The error name.

`.database1: string`
:   One of the databases.

`.database2: string`
:   The other database.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();

await db.connect("http://localhost:8000");
await db.signin({ user: "root", pass: "root" });
await Promise.all([
  db.use({ namespace: "baz" database: "foo" }),
  db.use({ namespace: "baz" database: "bar" }),
]);
```

## MissingNamespaceError

Inherits from: `SurrealError`

This error is thrown when a namespace is not selected before selecting a database. Or, it is thrown when attempting to switch to a state where a namespace is not selected while a database is still selected.

`.name: "MissingNamespaceError"`
:   The error name.

`.database: string`
:   The database name.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.use({ database: "example" }); // throws MissingNamespaceError: The namespace must be specified before the database.
```

## RpcResponseError

Inherits from: `SurrealError`

This error is thrown when an RPC response indicates an error. There are no problems with the communication or response body decoding according to the protocol to which you are connected, but it means that SurrealDB cannot process the RPC request.

`.name: "RpcResponseError"`
:   The error name.

`.id?: string`
:   ID that identifies the RPC request. The ID always starts with `<method name>_`.

`.code: number`
:   This is not clearly stated in the SurrealDB documentation, but it is probably the JSON-RPC error code.

## QueryFailedError

Inherits from: `SurrealAggregateError`

This error is thrown when a query fails.

`.name: "QueryFailedError"`
:   The error name.

`.cause: string[]`
:   List of errors.

For example, this error is thrown in the following case:

```ts
await using db = new Surreal();
await db.connect("http://localhost:8000");

await db.query("OUTPUT 'Hello'"); // throws QueryFailedError: Query failed with 1 error(s)
```

## Closed

Inherits from: `SurrealError`

This error indicates that the connection has been forcibly closed.

`.name: "Closed"`
:   The error name.

For example, you get this error in the following case:

```ts
const db = new Surreal();

db.on("<event_name>", ({ signal }) => {
  return new Promise((resolve, reject) => {
    signal.addEventListener("abort", function() {
      console.error(this.reason); // Closed
      reject();
    })

    // ...
  });
});

await db.close({ force: true });
```

## EngineError

Inherits from: `SurrealError`

This indicates that the error is due to the engine. It is sent via event listeners.

`.name: "EngineError"`
:   The error name.

`.fatal: boolean | undefined`
:   Indicates whether the error is fatal.

## HttpEngineError

Inherits from: `EngineError`

This indicates that the error is due to the HTTP engine. Currently it is only defined and not used.

`.name: "HttpEngineError"`
:   The error name.

## WebSocketEngineError

Inherits from: `EngineError`

This indicates that the error is due to the WebSocket engine. It is used when parsing the response body fails before finding the ID that identifies the RPC request, when an error event is received from the WebSocket object, when the WebSocket connection is closed, etc.

`.name: "WebSocketEngineError"`
:   The error name.

`.code: number`
:   One of the following status codes:
    - `1000`~`1015`: [MDN Reference](https://developer.mozilla.org/docs/Web/API/CloseEvent/code)
    - `3150`: Indicates that an "error" event was received from the `WebSocket`[^1] instance.
    - `3151`: Indicates that an error occurred within the "open" event handler registered by the WebSocket engine. In most cases, a `StateTransitionError` occurs during the transition from the connecting state to the connected state.
    - `3152`: Indicates that an error occurred within the "message" event handler registered by the WebSocket engine. There are various possible causes, but it is highly likely that a `ResponseError` or `RpcResponseError` was thrown within the event handler. If you trace back the origin, it is likely caused by requesting a SurrealQL custom query with incorrect syntax in the `.query()` method, or by requesting with invalid RPC parameters. In that case, the RPC caller method (by default) times out after 5 seconds and fails.
    - `3153`: Indicates that the ping failed to be sent or received. It may be a temporary error, but if you keep receiving it, the connection may not be maintained.
    - `3154`: Indicates that an error occurred within the "close" event handler registered by the WebSocket engine. In most cases, a `StateTransitionError` occurs during the transition from the disconnecting state to the disconnected state.
:   In addition, the following status codes are excluded from the above. This SDK does not treat them as errors.
    - `1000`: Successfully disconnected.
    - `1001`: Early disconnections are common.
    - `1004`: Reserved.
    - `1005`: Reserved.
    - `1006`: Reserved.
    - `1015`: Reserved.

^1: `WebSocket` may be a class defined in the runtime's global variable, or a class from the `ws` SDK.

## StateTransitionError

Inherits from: `SurrealAggregateError`

This error is thrown when the execution of an event listener fails during a state transition.

`.name: "StateTransitionError"`
:   The error name.

`.from: string`
:   The state when the state transition started.

`.to: string`
:   The destination state for the transition.

`.fallback: string`
:   Alternative transition destination if the state transition fails. If it is the same as `.to`, it means the transition was forced.

For example, this error is thrown in the following case:

```ts
import { Surreal } from "@tai-kun/surrealdb";
import { CONNECTING } from "@tai-kun/surrealdb/engine";

await using db = new Surreal();

db.on(CONNECTING, () => {
  throw new Error("Don't let me connect !!!")
});

await db.connect("http://localhost:8000"); // throws ...
// StateTransitionError: The transition from `3` to `0` failed, falling back to `3`.
//     at <stack trace ...> {
//   cause: [
//     Error: Don't let me connect !!!
//         at <stack trace ...>
//   ],
//   from: 3,
//   to: 0,
//   fallback: 3
// }
```

## ConnectionUnavailableError

Inherits from: `SurrealError`

This error is thrown when attempting to send an RPC request while not connected.

`.name: "ConnectionUnavailableError"`
:   The error name.

## ResponseError

Inherits from: `SurrealError`

This error is thrown when the response cannot be parsed as a PRC response. It is different from `RpcResponseError`.

`.name: "ResponseError"`
:   The error name.

## CborError

Inherits from: `SurrealError`

This is the class that all CBOR-related errors explicitly thrown by `@tai-kun/surrealdb/cbor` inherit from. It is not directly thrown.

`.name: "CborError"`
:   The error name.

## CborWellFormednessError

Inherits from: `CborError`

This is the class that all CBOR decoding-related errors explicitly thrown by `@tai-kun/surrealdb/cbor` inherit from. It is not directly thrown.

Refer to Appendix F "Well-Formedness Errors and Examples" in RFC8949 [https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-](https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-)

`.name: "CborWellFormednessError"`
:   The error name.

## CborTooMuchDataError

Inherits from: `CborWellFormednessError`

This indicates that there are unconsumed input bytes remaining.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2)

`.name: "CborTooMuchDataError"`
:   The error name.

## CborTooLittleDataError

Inherits from: `CborWellFormednessError`

This indicates that the input bytes are incomplete.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4)

`.name: "CborTooLittleDataError"`
:   The error name.

## CborSyntaxError

Inherits from: `CborWellFormednessError`

This indicates that the input bytes do not match the CBOR encoding format.

[RFC8949](https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6)

`.name: "CborSyntaxError"`
:   The error name.

## CborMaxDepthReachedError

Inherits from: `CborError`

This error is thrown when the depth of the JavaScript object reaches the maximum value during CBOR encoding or decoding. The depth increases by 1 each time you enter an object or array.

`.name: "CborMaxDepthReachedError"`
:   The error name.

`.maxDepth: number`
:   The maximum depth.

## CborUnsafeMapKeyError

Inherits from: `CborError`

This error is thrown when an unsafe map key is found during CBOR encoding or decoding.

`.name: "CborUnsafeMapKeyError"`
:   The error name.

`.key: unknown`
:   The map key that was determined to be unsafe.
