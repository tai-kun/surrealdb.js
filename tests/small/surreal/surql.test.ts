import { Slot, surql } from "@tai-kun/surrealdb";
import type { Writer } from "@tai-kun/surrealdb/cbor";
import { SurrealTypeError } from "@tai-kun/surrealdb/errors";
import { EncodedCBOR } from "@tai-kun/surrealdb/formatter";
import { afterEach, describe, expect, test, vi } from "vitest";

describe(".text はエンコード済み", () => {
  const CreateUserQuery = surql`CREATE user:foo;`;

  const writeBytes = vi.fn();
  const writer = { writeBytes } as unknown as Writer;

  afterEach(() => {
    writeBytes.mockClear();
  });

  test(".text", () => {
    expect(CreateUserQuery.text).toBeInstanceOf(EncodedCBOR);
    (CreateUserQuery.text as EncodedCBOR).toCBOR(writer);
    expect(writeBytes.mock.calls).toStrictEqual([
      [
        // `CREATE user:foo;`
        new Uint8Array([
          0x70, // mt: 3, ai: 16
          0x43, // `C`
          0x52, // `R`
          0x45, // `E`
          0x41, // `A`
          0x54, // `T`
          0x45, // `E`
          0x20, // ` `
          0x75, // `u`
          0x73, // `s`
          0x65, // `e`
          0x72, // `r`
          0x3a, // `:`
          0x66, // `f`
          0x6f, // `o`
          0x6f, // `o`
          0x3b, // `;`
        ]),
      ],
    ]);
  });
});

describe("変数を埋め込める", () => {
  const variable = 10;
  const CreateUserQuery = surql`RETURN ${variable};`;

  const writeBytes = vi.fn();
  const writer = { writeBytes } as unknown as Writer;

  afterEach(() => {
    writeBytes.mockClear();
  });

  test(".text", () => {
    expect(CreateUserQuery.text).toBeInstanceOf(EncodedCBOR);
    (CreateUserQuery.text as EncodedCBOR).toCBOR(writer);
    expect(writeBytes.mock.calls).toStrictEqual([
      [
        // `RETURN $_jst_0;`
        new Uint8Array([
          0x6f, // mt: 3, ai: 15
          0x52, // `R`
          0x45, // `E`
          0x54, // `T`
          0x55, // `U`
          0x52, // `R`
          0x4e, // `N`
          0x20, // ` `
          0x24, // `$`
          0x5f, // `_`
          0x6a, // `j`
          0x73, // `s`
          0x74, // `t`
          0x5f, // `_`
          0x30, // `0`
          0x3b, // `;`
        ]),
      ],
    ]);
  });

  test(".vars", () => {
    expect(CreateUserQuery.vars).toHaveProperty("_jst_0");

    expect(CreateUserQuery.vars["_jst_0"]).toBeInstanceOf(EncodedCBOR);
    (CreateUserQuery.vars["_jst_0"] as EncodedCBOR).toCBOR(writer);
    expect(writeBytes.mock.calls).toStrictEqual([
      [
        // `10`
        new Uint8Array([
          0x0a, // mt: 0, ai: 10
        ]),
      ],
    ]);
  });
});

describe("スロットを埋め込める", () => {
  const ValueSlot = surql.slot("value");
  const CreateUserQuery = surql`RETURN ${ValueSlot};`;

  const writeBytes = vi.fn();
  const writer = { writeBytes } as unknown as Writer;

  afterEach(() => {
    writeBytes.mockClear();
  });

  test(".text", () => {
    expect(CreateUserQuery.text).toBeInstanceOf(EncodedCBOR);
    (CreateUserQuery.text as EncodedCBOR).toCBOR(writer);
    expect(writeBytes.mock.calls).toStrictEqual([
      [
        // `RETURN $value;`
        new Uint8Array([
          0x6e, // mt: 3, ai: 14
          0x52, // `R`
          0x45, // `E`
          0x54, // `T`
          0x55, // `U`
          0x52, // `R`
          0x4e, // `N`
          0x20, // ` `
          0x24, // `$`
          0x76, // `v`
          0x61, // `a`
          0x6c, // `l`
          0x75, // `u`
          0x65, // `e`
          0x3b, // `;`
        ]),
      ],
    ]);
  });

  test(".vars", () => {
    expect(CreateUserQuery.vars).not.toHaveProperty("value");
  });

  test(".slots", () => {
    expect(CreateUserQuery.slots).toMatchObject([
      new Slot("value", true),
    ]);
  });
});

describe("同じオブジェクトを 2 度以上処理しない", () => {
  const variable = [10];
  // const ValueSlot = surql.slot("value");
  const CreateUserQuery = surql`RETURN [${variable}, ${variable}];`;

  const writeBytes = vi.fn();
  const writer = { writeBytes } as unknown as Writer;

  afterEach(() => {
    writeBytes.mockClear();
  });

  test(".text", () => {
    expect(CreateUserQuery.text).toBeInstanceOf(EncodedCBOR);
    (CreateUserQuery.text as EncodedCBOR).toCBOR(writer);
    expect(writeBytes.mock.calls).toStrictEqual([
      [
        // `RETURN [$_jst_0, $_jst_0];`
        new Uint8Array([
          0x78, // mt: 3, ai: 24
          0x1a, // 26
          0x52, // `R`
          0x45, // `E`
          0x54, // `T`
          0x55, // `U`
          0x52, // `R`
          0x4e, // `N`
          0x20, // ` `
          0x5b, // `[`
          0x24, // `$`
          0x5f, // `_`
          0x6a, // `j`
          0x73, // `s`
          0x74, // `t`
          0x5f, // `_`
          0x30, // `0`
          0x2c, // `,`
          0x20, // ` `
          0x24, // `$`
          0x5f, // `_`
          0x6a, // `j`
          0x73, // `s`
          0x74, // `t`
          0x5f, // `_`
          0x30, // `0`
          0x5d, // `]`
          0x3b, // `;`
        ]),
      ],
    ]);
  });

  test(".vars", () => {
    expect(CreateUserQuery.vars).toHaveProperty("_jst_0");
    expect(CreateUserQuery.vars).not.toHaveProperty("_jst_1");

    expect(CreateUserQuery.vars["_jst_0"]).toBeInstanceOf(EncodedCBOR);
    (CreateUserQuery.vars["_jst_0"] as EncodedCBOR).toCBOR(writer);
    expect(writeBytes.mock.calls).toStrictEqual([
      [
        // `[10]`
        new Uint8Array([
          0x81, // mt: 4, ai: 1
          0x0a, // 10
        ]),
      ],
    ]);
  });
});

describe("エラー", () => {
  test("特定の変数名から始まるスロットでエラー", () => {
    const ValueSlot = surql.slot("_jst_foo");

    expect(() => surql`RETURN ${ValueSlot};`).toThrowErrorMatchingSnapshot();
  });
});
