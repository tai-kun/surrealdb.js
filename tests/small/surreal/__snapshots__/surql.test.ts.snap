// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`エラー > 特定の変数名から始まるスロットでエラー 1`] = `
{
  "actual": "_jst_foo",
  "expected": "a variable name that do not start with "_jst_"",
  "message": "Expected a variable name that do not start with "_jst_" but got type String of value "_jst_foo".",
  "name": "SurrealValueError",
}
`;
